### PR TITLE
OLS-651: additional cert for app server

### PR DESCRIPTION
**Pull request by raptorsun**
## Description

work in progress, do not merge

## Type of change

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Configuration Update
- [ ] Bump-up dependent library

## Related Tickets & Documents

- Related Issue # [OLS-651](https://issues.redhat.com//browse/OLS-651), OLS-650, [OLS-385](https://issues.redhat.com//browse/OLS-385)
- Closes # [OLS-651](https://issues.redhat.com//browse/OLS-651)

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] PR has passed all pre-merge test jobs.
- [ ] If it is a core feature, I have added thorough tests.

## Testing
- Please provide detailed steps to perform tests related to this code change.
- How were the fix/results from this change verified? Please provide relevant screenshots or results.

### CODE REVIEW

File: api/v1alpha1/olsconfig_types.go [hunk 1]
```diff
@@ -104,6 +104,9 @@ type OLSSpec struct {
 	// User data collection switches
 	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="User Data Collection"
 	UserDataCollection UserDataCollectionSpec `json:"userDataCollection,omitempty"`
+	// Additional CA for TLS communication
```
Comment (xrajesh) [inline api/v1alpha1/olsconfig_types.go:107]: `Additional CA certificates for TLS communication between OLS service and LLM Provider`  @raptorsun little more clear I believe.

File: internal/controller/constants.go [hunk 2]
```diff
@@ -90,6 +94,8 @@ const (
 	LLMProviderHashStateCacheKey = "llmprovider-hash"
 	// AzureOpenAIType is the name of the Azure OpenAI provider type
 	AzureOpenAIType = "azure_openai"
+	// AdditionalCAHashStateCacheKey is the key of the hash value of the additional CA certificates
```
Comment (xrajesh) [inline internal/controller/constants.go:97]: @raptorsun Line 73 is also saying the same.

**Comment by xrajesh** [comment]
@raptorsun IF writing to the certifi dorectory is a problem are we able to update the permissions - something like this - I have not tested it - trying a PR like this github.com/openshift/lightspeed-service/pull/1525

**Comment by raptorsun** [comment]
> @raptorsun IF writing to the certifi dorectory is a problem are we able to update the permissions - something like this - I have not tested it - trying a PR like this github.com/[openshift/lightspeed-service/pull/1525](https://github.com/openshift/lightspeed-service/pull/1525)

checking with new image from https://github.com/openshift/lightspeed-service/pull/1533
will post result soon

**Comment by raptorsun** [comment]
After testing with the image from https://github.com/openshift/lightspeed-service/pull/1533
 , I find it successfully makes the certifi file writable in the pod.
On the access settings, I see 2 solutions:
1. explicitly set a RunAsUser (for example 1000600000) in the pod template spec with setting the owner of the certifi cert file to that UID (1000600000).
2. Make the file writable to anyone (666) without modifying the runner UID in the pod spec.

The first solution seems more secure. @xrajesh

File: internal/controller/ols_app_server_deployment.go [hunk 3]
```diff
@@ -336,9 +374,11 @@ func (r *OLSConfigReconciler) updateOLSDeployment(ctx context.Context, existingD
 	}
 
 	// validate container specs
-	if !containersEqual(existingDeployment.Spec.Template.Spec.Containers, desiredDeployment.Spec.Template.Spec.Containers) {
+	if !containersEqual(existingDeployment.Spec.Template.Spec.Containers, desiredDeployment.Spec.Template.Spec.Containers) ||
+		!containersEqual(existingDeployment.Spec.Template.Spec.InitContainers, desiredDeployment.Spec.Template.Spec.InitContainers) {
```
Comment (xrajesh) [inline internal/controller/ols_app_server_deployment.go:378]: @raptorsun We dont need the check for initContainers anymore.

File: internal/controller/ols_app_server_deployment.go [hunk 4]
```diff
@@ -336,9 +374,11 @@ func (r *OLSConfigReconciler) updateOLSDeployment(ctx context.Context, existingD
...
 		changed = true
 		existingDeployment.Spec.Template.Spec.Containers = desiredDeployment.Spec.Template.Spec.Containers
+		existingDeployment.Spec.Template.Spec.InitContainers = desiredDeployment.Spec.Template.Spec.InitContainers
```
Comment (xrajesh) [inline internal/controller/ols_app_server_deployment.go:381]: @raptorsun We dont need the check for initContainers anymore.

File: internal/controller/utils.go [hunk 5]
```diff
@@ -250,7 +252,8 @@ func deploymentSpecEqual(a, b *appsv1.DeploymentSpec) bool {
 		return false
 	}
 
-	return containersEqual(a.Template.Spec.Containers, b.Template.Spec.Containers)
+	return containersEqual(a.Template.Spec.Containers, b.Template.Spec.Containers) &&
+		containersEqual(a.Template.Spec.InitContainers, b.Template.Spec.InitContainers)
```
Comment (xrajesh) [inline internal/controller/utils.go:256]: @raptorsun We dont need the check for initContainers anymore.

**Comment by raptorsun** [comment]
@xrajesh thank you for the review. all comment have been addressed now :)

File: internal/controller/ols_app_server_deployment.go [hunk 6]
```diff
@@ -274,17 +310,19 @@ func (r *OLSConfigReconciler) updateOLSDeployment(ctx context.Context, existingD
 		existingDeployment.Annotations[OLSAppTLSHashKey] != r.stateCache[OLSAppTLSHashStateCacheKey] ||
 		existingDeployment.Annotations[LLMProviderHashKey] != r.stateCache[LLMProviderHashStateCacheKey] {
```
Comment (xrajesh) [inline internal/controller/ols_app_server_deployment.go:311]: @raptorsun If just the AdditionalCAHashStateCacheKey changes ( direct update in the additional CR configmap and  no config CR change, no AppTLS change ,no Provider change  ) , that will trigger a reconcile . During that case  , if deployment exists and we do  updateOLSDeployment - we want to update the annotations with  new AdditionalCAHashStateCacheKey right ? are we missing a case for that ?

**Comment by raptorsun** [comment]
/override "Red Hat Konflux / ols-enterprise-contract / bundle"
/override "Red Hat Konflux / ols-enterprise-contract / lightspeed-operator"
/override "Red Hat Konflux / ols-enterprise-contract / test-bundle"

**Pull request merged**